programs/mango-v4/tests/test_reduce_only.rs

```diff
@@ -0,0 +1,571 @@
+use mango_v4::state::{Bank, MangoAccount, PerpMarket, Side};
+use program_test::*;
+use solana_program_test::*;
+use solana_sdk::transport::TransportError;
+
+mod program_test;
+
+#[tokio::test]
+async fn test_reduce_only_token() -> Result<(), TransportError> {
+    let context = TestContext::new().await;
+    let solana = &context.solana.clone();
+
+    let admin = TestKeypair::new();
+    let owner = context.users[0].key;
+    let payer = context.users[1].key;
+    let mints = &context.mints[0..2];
+    let payer_mint_accounts = &context.users[1].token_accounts[0..=2];
+
+    let initial_token_deposit = 10_000;
+
+    //
+    // SETUP: Create a group and an account
+    //
+
+    let GroupWithTokens { group, tokens, .. } = GroupWithTokensConfig {
+        admin,
+        payer,
+        mints: mints.to_vec(),
+        ..GroupWithTokensConfig::default()
+    }
+    .create(solana)
+    .await;
+
+    //
+    // SETUP: Prepare accounts
+    //
+    let account_0 = create_funded_account(
+        &solana,
+        group,
+        owner,
+        0,
+        &context.users[1],
+        &mints[0..2],
+        initial_token_deposit,
+        0,
+    )
+    .await;
+
+    // make token reduce only
+    send_tx(
+        solana,
+        TokenMakeReduceOnly {
+            admin,
+            group,
+            mint: mints[0].pubkey,
+        },
+    )
+    .await
+    .unwrap();
+
+    // deposit without reduce_only should fail
+    let res = send_tx(
+        solana,
+        TokenDepositInstruction {
+            amount: 10,
+            reduce_only: false,
+            account: account_0,
+            owner,
+            token_account: payer_mint_accounts[0],
+            token_authority: payer,
+            bank_index: 0,
+        },
+    )
+    .await;
+    assert!(res.is_err());
+
+    // deposit with reduce_only should pass silently
+    send_tx(
+        solana,
+        TokenDepositInstruction {
+            amount: 10,
+            reduce_only: true,
+            account: account_0,
+            owner,
+            token_account: payer_mint_accounts[0],
+            token_authority: payer,
+            bank_index: 0,
+        },
+    )
+    .await
+    .unwrap();
+    let mango_account_0 = solana.get_account::<MangoAccount>(account_0).await;
+    let bank = solana.get_account::<Bank>(tokens[0].bank).await;
+    let native = mango_account_0.tokens[0].native(&bank);
+    assert_eq!(native.to_num::<u64>(), initial_token_deposit);
+
+    // withdraw all should pass
+    send_tx(
+        solana,
+        TokenWithdrawInstruction {
+            amount: initial_token_deposit,
+            allow_borrow: false,
+            account: account_0,
+            owner,
+            token_account: payer_mint_accounts[0],
+            bank_index: 0,
+        },
+    )
+    .await
+    .unwrap();
+
+    // borrowing should fail
+    let res = send_tx(
+        solana,
+        TokenWithdrawInstruction {
+            amount: initial_token_deposit,
```

Comment:
maybe just borrow 1 - possibly there isn't enough collateral to borrow this much